File: SwiftAlgorithm/BOJ/스택,큐/1874.swift
//
//  main.swift
//  SwiftAlgorithm
//
//  Created by 유정주 on 2022/03/03.
//

//1874 스택 수열
import Foundation

let count = Int(readLine()!)!
var inputs: [Int] = []
var stack: [Int] = []
var result: [String] = []

for _ in 0..<count {
    let num = Int(readLine()!)!
    inputs.append(num)
}
inputs.reverse()

for i in 1...count {
    if let top = stack.last, top == inputs.last! {
        result.append("-")
        stack.removeLast()
    } else {
        result.append("+")
        stack.append(i)
        
        while !stack.isEmpty {
            if let top = stack.last, top == inputs.last! {
                result.append("-")
                stack.removeLast()
                inputs.removeLast()
            } else {
                break
            }
        }
    }
}

if inputs.isEmpty {
    for s in result {
        print("\(s)")
    }
} else {
    print("NO")
}

File: Python/BOJ/스택/1874.py
import sys

n = int(sys.stdin.readline().rstrip())
inputs = []
for _ in range(n):
  inputs.append(int(sys.stdin.readline().rstrip()))

inputs = inputs[::-1]

stack = []
result = []
for i in range(1, n+1):
  result += "+"
  stack.append(i)
  while stack:
    if stack[-1] == inputs[-1]:
      result += "-"
      stack.pop()
      inputs.pop()
    else:
      break

if not inputs:
  for op in result:
    print(op)
else:
  print("NO")
